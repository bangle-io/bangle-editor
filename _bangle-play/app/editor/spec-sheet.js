import '@banglejs/core/style.css';
import '@banglejs/tooltip/style.css';

import * as collab from '@banglejs/collab/client/collab-extension';
import { emoji } from '@banglejs/emoji/index';
import '@banglejs/emoji/emoji.css';
import '@banglejs/react-menu/style.css';
import '@banglejs/markdown-front-matter/style.css';
import './extensions-override.css';
import { trailingNode } from '@banglejs/trailing-node';
import { timestamp } from '@banglejs/timestamp';
import { SpecRegistry } from '@banglejs/core/spec-registry';
import stopwatch from '@banglejs/react-stopwatch';
import sticker from '@banglejs/react-sticker';
import { emojiSuggestMenu } from '@banglejs/react-menu';
import { markdownFrontMatter } from '@banglejs/markdown-front-matter';
import {
  bold,
  code,
  italic,
  strike,
  link,
  underline,
  doc,
  text,
  paragraph,
  blockquote,
  bulletList,
  codeBlock,
  hardBreak,
  heading,
  horizontalRule,
  listItem,
  orderedList,
  todoItem,
  todoList,
  image,
} from '@banglejs/core/components/index';

export const specRegistry = new SpecRegistry([
  doc.spec({ content: 'frontMatter? block+' }),
  text.spec(),
  paragraph.spec(),
  blockquote.spec(),
  bulletList.spec(),
  codeBlock.spec(),
  hardBreak.spec(),
  heading.spec(),
  horizontalRule.spec(),
  listItem.spec(),
  orderedList.spec(),
  todoItem.spec(),
  todoList.spec(),
  image.spec(),
  bold.spec(),
  code.spec(),
  italic.spec(),
  strike.spec(),
  link.spec(),
  underline.spec(),
  collab.spec(),
  emoji.spec(),
  emojiSuggestMenu.spec(),
  stopwatch.spec(),
  trailingNode.spec(),
  timestamp.spec(),
  sticker.spec(),
  markdownFrontMatter.spec(),
]);
