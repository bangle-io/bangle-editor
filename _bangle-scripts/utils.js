const path = require('path');
const globby = require('globby');
const fs = require('fs/promises');
const rootPath = path.resolve(__dirname, '..');

module.exports = { getPackages, mapPackages };

async function getPackages({ filter = 'all' } = {}) {
  const globResults = await globby([path.join(rootPath, '**/package.json')]);
  let results = await Promise.all(
    globResults.map(async (path) => [
      path,
      JSON.parse(await fs.readFile(path, 'utf-8')),
    ]),
  );

  results = results.filter(([filePath, packageObj]) => {
    if (filter === 'private') {
      return packageObj.private;
    }
    if (filter === 'public') {
      return !packageObj.private;
    }

    return true;
  });
  console.log(results);
  return results;
}

// cb - ([packagePath, packageObj]) => packageObj
async function mapPackages(cb, { filter } = {}) {
  const result = (await getPackages({ filter })).map((r) => {
    return [r[0], cb(r)];
  });

  console.log(result);
  await Promise.all(
    result.map(([packagePath, packageObj]) => {
      return fs.writeFile(
        packagePath,
        JSON.stringify(packageObj, null, 2) + '\n',
        'utf-8',
      );
    }),
  );
}
