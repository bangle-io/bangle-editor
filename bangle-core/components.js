import { coreMarkPlugins, coreMarkSpec } from './marks/index';
import { coreNodeSpec, coreNodePlugins } from './nodes/index';
import * as history from './extensions/history';

export * from './nodes/index';
export * from './marks/index';
export * as history from './extensions/history';

export function coreSpec(opts = {}) {
  return [
    ...coreMarkSpec(opts.node),
    ...coreNodeSpec(opts.node),
    history.spec(opts.history),
  ];
}

export function corePlugins(opts = {}) {
  return [
    ...coreMarkPlugins(opts.node),
    ...coreNodePlugins(opts.node),
    history.plugins(opts.history),
  ];
}
