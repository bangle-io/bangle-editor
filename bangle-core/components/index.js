// marks
import * as bold from './bold';
import * as code from './code';
import * as italic from './italic';
import * as link from './link';
import * as strike from './strike';
import * as underline from './underline';

// nodes
import * as doc from './doc';
import * as paragraph from './paragraph';
import * as text from './text';
import * as blockquote from './blockquote';
import * as bulletList from './bullet-list';
import * as codeBlock from './code-block';
import * as hardBreak from './hard-break';
import * as heading from './heading';
import * as horizontalRule from './horizontal-rule';
import * as listItem from './list-item/list-item';
import * as orderedList from './ordered-list';
import * as todoItem from './todo-item';
import * as todoList from './todo-list';
import * as image from './image';

// components
import * as history from './history';

// marks
export * as bold from './bold';
export * as code from './code';
export * as italic from './italic';
export * as link from './link';
export * as strike from './strike';
export * as underline from './underline';

// nodes
export * as doc from './doc';
export * as paragraph from './paragraph';
export * as text from './text';
export * as blockquote from './blockquote';
export * as bulletList from './bullet-list';
export * as codeBlock from './code-block';
export * as hardBreak from './hard-break';
export * as heading from './heading';
export * as horizontalRule from './horizontal-rule';
export * as listItem from './list-item/list-item';
export * as orderedList from './ordered-list';
export * as todoItem from './todo-item';
export * as todoList from './todo-list';
export * as image from './image';

// components
export * as history from './history';

export function coreSpec(opts = {}) {
  return [...coreMarkSpec(opts), ...coreNodeSpec(opts)];
}

export function corePlugins(opts = {}) {
  return [
    ...coreMarkPlugins(opts),
    ...coreNodePlugins(opts),
    history.plugins(opts.history),
  ];
}

function coreMarkSpec(options = {}) {
  return [
    bold.spec(options.bold),
    code.spec(options.code),
    italic.spec(options.italic),
    strike.spec(options.strike),
    link.spec(options.link),
    underline.spec(options.underline),
  ];
}

function coreMarkPlugins(options = {}) {
  return [
    bold.plugins(options.bold),
    code.plugins(options.code),
    italic.plugins(options.italic),
    strike.plugins(options.strike),
    link.plugins(options.link),
    underline.plugins(options.underline),
  ];
}

function coreNodeSpec(options = {}) {
  return [
    doc.spec(options.doc),
    text.spec(options.text),
    paragraph.spec(options.paragraph),
    blockquote.spec(options.blockquote),
    bulletList.spec(options.bulletList),
    codeBlock.spec(options.codeBlock),
    hardBreak.spec(options.hardBreak),
    heading.spec(options.heading),
    horizontalRule.spec(options.horizontalRule),
    listItem.spec(options.listItem),
    orderedList.spec(options.orderedList),
    todoItem.spec(options.todoItem),
    todoList.spec(options.todoList),
    image.spec(options.image),
  ];
}

function coreNodePlugins(options = {}) {
  return [
    paragraph.plugins(options.paragraph),
    blockquote.plugins(options.blockquote),
    bulletList.plugins(options.bulletList),
    codeBlock.plugins(options.codeBlock),
    hardBreak.plugins(options.hardBreak),
    heading.plugins(options.heading),
    horizontalRule.plugins(options.horizontalRule),
    listItem.plugins(options.listItem),
    orderedList.plugins(options.orderedList),
    todoItem.plugins(options.todoItem),
    todoList.plugins(options.todoList),
    image.plugins(options.image),
  ];
}
