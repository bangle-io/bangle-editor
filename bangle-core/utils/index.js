export * from './dom-serialization-helpers';
export * from './keymap';
