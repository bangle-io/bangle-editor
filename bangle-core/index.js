import * as logging from './utils/logging';
import browser, { isChromeWithSelectionBug } from './utils/browser';
export * from './components/index';
export * from './editor';
export { logging, browser, isChromeWithSelectionBug };
export * from './pm-plugin';
export * from './node-view';
export * from './pm-model';
