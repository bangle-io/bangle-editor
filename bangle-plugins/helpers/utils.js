import { bangleWarn } from 'bangle-core/utils/js-utils';
import { PluginKey } from 'prosemirror-state';

export function pluginKeyStore() {
  const store = new WeakMap();
  return {
    get(parentKey, childKeyId) {
      validPluginKey(parentKey);
      const family = store.get(parentKey);
      if (!family || !family[childKeyId] || !childKeyId) {
        bangleWarn(
          `There is no key named ${childKeyId} associated with the pluginKey =`,
          parentKey,
          `\nPlease make sure you are passing the pluginKey correctly to your components.`,
        );

        throw new Error(`Cannot find the pluginKey '${childKeyId}'`);
      }

      return store.get(parentKey)[childKeyId];
    },

    create(parentKey, childKeyId) {
      validPluginKey(parentKey);
      const key = new PluginKey(childKeyId);
      if (store.has(parentKey)) {
        const children = store.get(parentKey);
        if (children[childKeyId]) {
          bangleWarn(
            `There is already a key ${childKeyId} associated with the pluginKey =`,
            parentKey,
          );

          throw new Error(
            `Parent is already associated with key'${childKeyId}'`,
          );
        }
        children[childKeyId] = key;
      } else {
        store.set(parentKey, { [childKeyId]: key });
      }
      return key;
    },
  };
}

function validPluginKey(key) {
  if (!(key instanceof PluginKey)) {
    bangleWarn('The key is not a valid pluginKey', key);
    throw new Error(`Please pass a valid plugin key`);
  }
}
