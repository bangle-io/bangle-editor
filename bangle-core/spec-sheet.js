import { Schema } from 'prosemirror-model';
import { coreSpec } from './components/index';
import { bangleWarn, recursiveFlat } from './utils/js-utils';

export class SpecSheet {
  constructor(rawSpecs = coreSpec()) {
    this._spec = recursiveFlat(rawSpecs);

    this._spec.forEach(validateSpec);

    if (this._spec.length !== new Set(this._spec.map((r) => r.name)).size) {
      bangleWarn(
        'The specSheet has one or more specs with the same name',
        this._spec,
      );
      throw new Error('Duplicate spec error, please check your specSheet');
    }
    this._schema = createSchema(this._spec);
    this._options = Object.fromEntries(
      this._spec
        .filter((spec) => spec.options)
        .map((spec) => [spec.name, spec.options]),
    );
  }

  get spec() {
    return this._spec;
  }

  get schema() {
    return this._schema;
  }

  get options() {
    return this._options;
  }
}

function createSchema(specSheet) {
  let nodes = [];
  let marks = [];
  let topNode;
  for (const spec of specSheet) {
    if (spec.type === 'node') {
      nodes.push([spec.name, spec.schema]);
    } else if (spec.type === 'mark') {
      marks.push([spec.name, spec.schema]);
    } else if (spec.type === 'component') {
    } else {
      throw new Error(spec.name + ' unknown type: ' + spec.type);
    }
    if (spec.topNode === true) {
      topNode = spec.name;
    }
  }

  return new Schema({
    topNode,
    nodes: Object.fromEntries(nodes),
    marks: Object.fromEntries(marks),
  });
}

function validateSpec(spec) {
  if (!spec.name) {
    bangleWarn("The spec didn't have a name field", spec);
    throw new Error('Spec must have a name');
  }
  if (!['node', 'mark', 'component'].includes(spec.type)) {
    bangleWarn('The spec must be of type node, mark or component ', spec);
    throw new Error('Invalid spec type');
  }
  if (['node', 'mark'].includes(spec.type) && !spec.schema) {
    bangleWarn(
      "The spec of type 'mark' or 'node' must have a schema field",
      spec,
    );
    throw new Error('Invalid spec schema');
  }
}
