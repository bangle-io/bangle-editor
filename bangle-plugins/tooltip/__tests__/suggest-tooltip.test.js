/**
 * @jest-environment jsdom
 */

/** @jsx psx */

import {
  psx,
  typeText,
  sendKeyToPm,
  renderTestEditor,
} from 'bangle-core/test-helpers/index';

import { typeChar } from 'bangle-core/test-helpers/index';
import { PluginKey, Selection } from 'prosemirror-state';
import { corePlugins, coreSpec } from 'bangle-core/components';
import { SpecSheet } from 'bangle-core/spec-sheet';
import { sleep } from 'bangle-core/utils/js-utils';
import { suggestTooltip } from '../index';
// We are using char code to differentiate between different schema
// 47 is char code for '/'
const suggestTriggerMarkSlash = (content) => (
  <suggest_slash trigger="/">{content}</suggest_slash>
);

describe('suggest basic show and hide', () => {
  let suggestionKey = new PluginKey(),
    testEditor;

  const specSheet = new SpecSheet([
    ...coreSpec(),
    suggestTooltip.spec({ markName: 'suggest_slash', trigger: '/' }),
  ]);
  const plugins = [
    ...corePlugins(),
    suggestTooltip.plugins({
      key: suggestionKey,
      markName: 'suggest_slash',
      trigger: '/',
    }),
  ];
  const getTooltipState = (state) => {
    return suggestionKey.getState(state);
  };

  beforeEach(async () => {
    testEditor = renderTestEditor({ specSheet, plugins });
  });

  test('when no trigger', async () => {
    const { view } = await testEditor(
      <doc>
        <para>foo[]bar</para>
      </doc>,
    );

    expect(view.dom.parentNode).toMatchSnapshot();
  });

  test.each([
    <doc>
      <para>[]</para>
    </doc>,
    <doc>
      <para>hello []</para>
    </doc>,
    <doc>
      <para>hello</para>
      <para>hello</para>
      <para>hello []</para>
    </doc>,
    <doc>
      <ul>
        <li>
          <para>hello []</para>
        </li>
      </ul>
    </doc>,
    <doc>
      <todoList>
        <todoItem>
          <para>hello []</para>
        </todoItem>
      </todoList>
    </doc>,
    <doc>
      <ol>
        <li>
          <para>hello []</para>
        </li>
      </ol>
    </doc>,
    <doc>
      <blockquote>
        <heading>hello []</heading>
      </blockquote>
    </doc>,
    <doc>
      <blockquote>
        <heading>hello</heading>
      </blockquote>
      <para>[]</para>
    </doc>,
  ])('Case %# different parent block', async (input) => {
    const { view } = await testEditor(input);

    expect(
      suggestTooltip.queryIsSuggestTooltipActive(suggestionKey)(view.state),
    ).toBe(false);
    typeChar(view, '/');
    typeText(view, 'check');
    expect(
      suggestTooltip.queryIsSuggestTooltipActive(suggestionKey)(view.state),
    ).toBe(true);
    expect(suggestTooltip.queryTriggerText(suggestionKey)(view.state)).toBe(
      'check',
    );
    expect(view.state.doc.toString()).toMatchSnapshot();

    expect(getTooltipState(view.state)).toEqual({
      markName: 'suggest_slash',
      show: true,
      trigger: '/',
    });
  });

  test('Not allowed in codeblock', async () => {
    const { view } = await testEditor(
      <doc>
        <ul>
          <li>
            <para>top</para>
          </li>
        </ul>
        <codeBlock>hello []</codeBlock>
      </doc>,
    );

    typeChar(view, '/');
    typeText(view, 'check');

    expect(view.state).toEqualDocAndSelection(
      <doc>
        <ul>
          <li>
            <para>top</para>
          </li>
        </ul>
        <codeBlock>hello /check[]</codeBlock>
      </doc>,
    );

    expect(getTooltipState(view.state)).toEqual({
      show: false,
      markName: 'suggest_slash',
      trigger: '/',
    });
  });

  test('trigger on an empty doc', async () => {
    const { view } = await testEditor(
      <doc>
        <para>[]</para>
      </doc>,
    );

    expect(
      suggestTooltip.queryIsSuggestTooltipActive(suggestionKey)(view.state),
    ).toBe(false);

    typeChar(view, '/');
    typeText(view, 'check');
    expect(
      suggestTooltip.queryIsSuggestTooltipActive(suggestionKey)(view.state),
    ).toBe(true);
    expect(suggestTooltip.queryTriggerText(suggestionKey)(view.state)).toBe(
      'check',
    );

    expect(view.state).toEqualDocAndSelection(
      <doc>
        <para>{suggestTriggerMarkSlash('/check')}[]</para>
      </doc>,
    );

    expect(getTooltipState(view.state)).toEqual({
      show: true,
      markName: 'suggest_slash',
      trigger: '/',
    });
  });

  test('types a trigger', async () => {
    const { view } = await testEditor(
      <doc>
        <para>foobar []</para>
      </doc>,
    );

    typeChar(view, '/');
    typeText(view, 'check');
    expect(suggestTooltip.queryTriggerText(suggestionKey)(view.state)).toBe(
      'check',
    );

    expect(view.state).toEqualDocAndSelection(
      <doc>
        <para>foobar {suggestTriggerMarkSlash('/check')}[]</para>
      </doc>,
    );

    expect(getTooltipState(view.state)).toEqual({
      show: true,
      markName: 'suggest_slash',
      trigger: '/',
    });

    // Give dom time to settle, as popper state updates as not sync
    await sleep(5);
    expect(view.dom.parentNode).toMatchSnapshot();
  });

  test('Selection going to other location hides the tooltip', async () => {
    const { view } = await testEditor(
      <doc>
        <para>[] hello</para>
      </doc>,
    );

    typeChar(view, '/');
    typeText(view, '');
    const { state } = view;

    // Outside the mark
    view.dispatch(state.tr.setSelection(Selection.atEnd(state.doc)));

    expect(getTooltipState(view.state)).toEqual({
      show: false,
      markName: 'suggest_slash',
      trigger: '/',
    });

    let tr = state.tr;

    // Inside the mark
    view.dispatch(tr.setSelection(Selection.near(tr.doc.resolve(2))));
    expect(getTooltipState(view.state)).toEqual({
      show: true,
      markName: 'suggest_slash',
      trigger: '/',
    });
    expect(
      suggestTooltip.queryIsSuggestTooltipActive(suggestionKey)(view.state),
    ).toBe(true);
  });

  test('Selection going to other location hides the tooltip', async () => {
    const { view } = await testEditor(
      <doc>
        <para>[] hello</para>
      </doc>,
    );

    typeChar(view, '/');
    typeText(view, '');
    const { state } = view;

    // Outside the mark
    view.dispatch(state.tr.setSelection(Selection.atEnd(state.doc)));

    expect(getTooltipState(view.state)).toEqual({
      show: false,
      markName: 'suggest_slash',
      trigger: '/',
    });

    let tr = state.tr;

    // Inside the mark
    view.dispatch(tr.setSelection(Selection.near(tr.doc.resolve(2))));
    expect(getTooltipState(view.state)).toEqual({
      show: true,
      markName: 'suggest_slash',
      trigger: '/',
    });
    expect(
      suggestTooltip.queryIsSuggestTooltipActive(suggestionKey)(view.state),
    ).toBe(true);
  });

  test('Query at end of document', async () => {
    const { view } = await testEditor(
      <doc>
        <para>[] hello</para>
      </doc>,
    );

    typeChar(view, '/');
    typeText(view, 'first');

    // Outside the mark
    view.dispatch(view.state.tr.setSelection(Selection.atEnd(view.state.doc)));

    typeText(view, ' ');
    typeChar(view, '/');
    typeText(view, 'second');

    expect(
      suggestTooltip.queryIsSuggestTooltipActive(suggestionKey)(view.state),
    ).toBe(true);
    expect(suggestTooltip.queryTriggerText(suggestionKey)(view.state)).toBe(
      'second',
    );

    let tr = view.state.tr;

    // Inside the mark
    view.dispatch(tr.setSelection(Selection.near(tr.doc.resolve(7))));
    expect(
      suggestTooltip.queryIsSuggestTooltipActive(suggestionKey)(view.state),
    ).toBe(true);
    expect(suggestTooltip.queryTriggerText(suggestionKey)(view.state)).toBe(
      'first',
    );
  });

  test('Really long query', async () => {
    const { view } = await testEditor(
      <doc>
        <para>[] hello</para>
      </doc>,
    );

    const longText =
      'firstfirstfirstfirst firstfirstfirstfirstfirst firstfirstfirstfirstfirstfirstfirstfirstfirstfirstfirstfirstfirstfirstfirstfirstfirstfirstfirst firstfirstfirstfirstfirstfirstfirstfirstfirstfirstfirstfirstfirstfirstfirst';
    typeChar(view, '/');
    typeText(view, longText);

    expect(
      suggestTooltip.queryIsSuggestTooltipActive(suggestionKey)(view.state),
    ).toBe(true);
    expect(suggestTooltip.queryTriggerText(suggestionKey)(view.state)).toBe(
      longText,
    );
  });

  test('Typing trigger text slowly', async () => {
    const { view } = await testEditor(
      <doc>
        <para>[] hello</para>
      </doc>,
    );

    typeChar(view, '/');
    typeText(view, 'c');
    typeText(view, 'h');
    typeText(view, 'e');
    typeText(view, 'c');
    typeText(view, 'k');

    expect(
      suggestTooltip.queryIsSuggestTooltipActive(suggestionKey)(view.state),
    ).toBe(true);
    expect(suggestTooltip.queryTriggerText(suggestionKey)(view.state)).toBe(
      'check',
    );
  });

  test.skip('When forward deleting the mark the tooltip hides', async () => {
    const { view } = await testEditor(
      <doc>
        <para>hello[]</para>
      </doc>,
    );

    // typeChar(view, '/');
    // typeText(view, 'check');
    // expect(suggestionsTooltip.queryIsSuggestTooltipActive(suggestionKey)(view.state)).toBe(true);

    // view.dispatch(view.state.tr.delete(1, 2));

    // view.dispatch(
    //   view.state.tr.setSelection(Selection.atStart(view.state.doc)),
    // );

    sendKeyToPm(view, 'Z');
    // sendKeyToPm(view, 'Backspace');
    // sendKeyToPm(view, 'Backspace');

    expect(view.state).toEqualDocAndSelection(
      <doc>
        <para>hello</para>
      </doc>,
    );
    expect(
      suggestTooltip.queryIsSuggestTooltipActive(suggestionKey)(view.state),
    ).toBe(true);
  });
});

describe('keybindings test', () => {
  let testEditor;
  let suggestionKey = new PluginKey();

  const opts = {
    key: suggestionKey,
    markName: 'suggest_slash',
    trigger: '/',
    placement: 'bottom-start',
    enterKeyName: 'Enter',
    arrowUpKeyName: 'ArrowUp',
    arrowDownKeyName: 'ArrowDown',
    escapeKeyName: 'Escape',
    // Use another key to mimic enter behaviour for example, Tab for entering
    alternateEnterKeyName: undefined,

    onUpdateTooltip: jest.fn(() => true),
    onHideTooltip: jest.fn(() => true),
    onEnter: jest.fn(() => true),
    onArrowDown: jest.fn(() => true),
    onArrowUp: jest.fn(() => true),
    onEscape: jest.fn(() => true),
  };

  const specSheet = new SpecSheet([
    ...coreSpec(),
    suggestTooltip.spec({ markName: 'suggest_slash', trigger: '/' }),
  ]);
  const plugins = [...corePlugins(), suggestTooltip.plugins(opts)];

  test('calls on* callbacks correctly', async () => {
    testEditor = renderTestEditor({ specSheet, plugins });
    const { view } = await testEditor(
      <doc>
        <para>[] foobar</para>
      </doc>,
    );

    typeChar(view, '/');
    typeText(view, 'check');
    expect(
      suggestTooltip.queryIsSuggestTooltipActive(suggestionKey)(view.state),
    ).toBe(true);
    expect(suggestTooltip.queryTriggerText(suggestionKey)(view.state)).toBe(
      'check',
    );

    expect(view.state).toEqualDocAndSelection(
      <doc>
        <para>{suggestTriggerMarkSlash('/check')}[] foobar</para>
      </doc>,
    );
    // since it is called many times
    expect(opts.onUpdateTooltip).toBeCalled();
    expect(opts.onHideTooltip).toBeCalledTimes(0);
    expect(
      suggestTooltip.queryIsSuggestTooltipActive(suggestionKey)(view.state),
    ).toBe(true);

    sendKeyToPm(view, 'Escape');
    expect(opts.onEscape).toBeCalledTimes(1);
    sendKeyToPm(view, 'Enter');
    expect(opts.onEnter).toBeCalledTimes(1);
    sendKeyToPm(view, 'ArrowUp');
    expect(opts.onArrowUp).toBeCalledTimes(1);
    sendKeyToPm(view, 'ArrowDown');
    expect(opts.onArrowDown).toBeCalledTimes(1);

    expect(
      suggestTooltip.queryIsSuggestTooltipActive(suggestionKey)(view.state),
    ).toBe(true);
    expect(suggestTooltip.queryTriggerText(suggestionKey)(view.state)).toBe(
      'check',
    );

    // Outside the mark
    view.dispatch(view.state.tr.setSelection(Selection.atEnd(view.state.doc)));
    // TODO for some mysterious reason this is called twice with the same state
    expect(opts.onHideTooltip).toBeCalledTimes(1);
  });
});

test.todo(
  'Check with differen trigger inline-suggest and that they dont interfere with each other',
);
