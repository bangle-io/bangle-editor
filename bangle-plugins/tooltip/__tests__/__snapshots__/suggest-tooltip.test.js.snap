// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`suggest basic show and hide Case 0 different parent block 1`] = `"doc(paragraph(suggest_slash(\\"/check\\")))"`;

exports[`suggest basic show and hide Case 1 different parent block 1`] = `"doc(paragraph(\\"hello \\", suggest_slash(\\"/check\\")))"`;

exports[`suggest basic show and hide Case 2 different parent block 1`] = `"doc(paragraph(\\"hello\\"), paragraph(\\"hello\\"), paragraph(\\"hello \\", suggest_slash(\\"/check\\")))"`;

exports[`suggest basic show and hide Case 3 different parent block 1`] = `"doc(bullet_list(list_item(paragraph(\\"hello \\", suggest_slash(\\"/check\\")))))"`;

exports[`suggest basic show and hide Case 4 different parent block 1`] = `"doc(todo_list(todo_item(paragraph(\\"hello \\", suggest_slash(\\"/check\\")))))"`;

exports[`suggest basic show and hide Case 5 different parent block 1`] = `"doc(ordered_list(list_item(paragraph(\\"hello \\", suggest_slash(\\"/check\\")))))"`;

exports[`suggest basic show and hide Case 6 different parent block 1`] = `"doc(blockquote(heading(\\"hello \\", suggest_slash(\\"/check\\"))))"`;

exports[`suggest basic show and hide Case 7 different parent block 1`] = `"doc(blockquote(heading(\\"hello\\")), paragraph(suggest_slash(\\"/check\\")))"`;

exports[`suggest basic show and hide types a trigger 1`] = `
<div
  data-testid="test-editor"
>
  <div
    class="ProseMirror bangle-editor content"
    contenteditable="true"
  >
    <p>
      foobar 
      <span
        data-bangle-name="suggest_slash"
        data-suggest-trigger="/"
        style="color: #0052CC"
      >
        /check
      </span>
    </p>
  </div>
  <div
    class="bangle-tooltip"
    data-popper-placement="bottom-start"
    data-show=""
    data-tooltip-name="suggest_slash__tooltipPlacementKey"
    role="tooltip"
    style="margin: 0px; position: absolute; left: 0px; top: 0px; transform: translate(0px, 0px);"
  >
    <div
      class="bangle-tooltip-content"
    >
      hello world
    </div>
  </div>
</div>
`;

exports[`suggest basic show and hide when no trigger 1`] = `
<div
  data-testid="test-editor"
>
  <div
    class="ProseMirror bangle-editor content"
    contenteditable="true"
  >
    <p>
      foobar
    </p>
  </div>
  <div
    class="bangle-tooltip"
    data-tooltip-name="suggest_slash__tooltipPlacementKey"
    role="tooltip"
  >
    <div
      class="bangle-tooltip-content"
    >
      hello world
    </div>
  </div>
</div>
`;
